### PR TITLE
Make Brave-specific options configurable through Windows Group Policy templates

### PR DESCRIPTION
**Pull request by bsclifton**
Updates Brave's group policies to match the style that Chromium does. 

Notable changes:
- Group policies have been removed from `src/brave/chromium_src/components/policy/tools/generate_policy_source.py` and broken out to individual files under `src/brave/components/policy/resources/templates/policy_definitions/BraveSoftware/`. This matches the style upstream in Chromium. For example, see https://source.chromium.org/chromium/chromium/src/+/main:components/policy/resources/templates/policy_definitions/Accessibility/AccessibilityShortcutsEnabled.yaml
- Group policies are now in a "group" called `BraveSoftware`.
- New hook added to `DEPS` file. This copies the Brave specific group policies found in `src/brave/components/policy/resources/templates/policy_definitions/` into the Chromium specific directory at ``src/components/policy/resources/templates/policy_definitions/`.
- `src/brave/chromium_src/components/policy/resources/policy_templates.py` was added to intercept the loading of the `policies.yaml` file upstream and add our own group policies.

This new approach generates a `policy_templates.zip` and `brave_policy_templates.zip` which has all of the Brave specific group policy values present.
- The `windows/examples/brave.reg` has examples for using all Brave specific registry keys
- The `windows/admx/brave.admx` group policy file has the Brave specific entries in it

image

image

This approach also does generate macOS (example plist file) and Linux (example JSON file) resources with the Brave specific group policy values- but we are not using those anywhere. More info about those can be found here:
https://source.chromium.org/chromium/chromium/src/+/main:components/policy/BUILD.gn

Fixes https://github.com/brave/brave-browser/issues/26502

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Building yourself
1. Be on Windows
2. `npm run build`
3. `npm run build -- --target=brave/components/policy:pack_policy_templates`

### Trying out the policy
1. Unzip `brave_policy_templates.zip` (in the out directory) to somewhere
2. Open `gpedit.msc` and right click `Administrative Templates` and pick `Add/Remove Templates`
3. Navigate into the directory you extracted `brave_policy_templates.zip` and add either the ADM or ADMX
4. Look for the group policies - you should see all of the Chromium ones and then the Brave specific ones are going to be under the `Brave Software settings`
image

### CODE REVIEW

File: build/util/update_group_policy.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at https://mozilla.org/MPL/2.0/.
+
+import os
+import shutil
+import hashlib
+
+# for running directly, you can uncomment and run
+import sys
+sys.path.append('C:\\bb3\\src\\brave\\script')
```
Comment (bsclifton) [inline build/util/update_group_policy.py:13]: whoops; need to remove this

File: patches/components-policy-resources-templates-policies.yaml.patch [hunk 2]
```diff
@@ -0,0 +1,19 @@
+diff --git a/components/policy/resources/templates/policies.yaml b/components/policy/resources/templates/policies.yaml
+index ecde8eaa2306e9fc46d4f2de636a1c3fbd1fd536..5081dcddac551a4bd3032d757c767202f8e45165 100644
+--- a/components/policy/resources/templates/policies.yaml
++++ b/components/policy/resources/templates/policies.yaml
+@@ -1293,6 +1293,14 @@ policies:
+   1292: AdHocCodeSigningForPWAsEnabled
+   1293: KioskWebAppOfflineEnabled
+   1294: GraduationEnablementStatus
++  1295: TorDisabled
```
Comment (bsclifton) [inline patches/components-policy-resources-templates-policies.yaml.patch:9]: I'm thinking about patching this with one entry:
`BRAVE_GROUP_POLICIES`

And then using a Chromium src override to replace that value with the ones in the `src/brave/components/policy/resources/templates/policy_definitions/BraveSoftware/` folder

Comment (bsclifton) [reply patches/components-policy-resources-templates-policies.yaml.patch:9]: actually- found a better place. We can intercept where policies.yaml gets loaded in `src/components/policy/resources/policy_templates.py` 😄 Stay tuned

Comment (bsclifton) [reply patches/components-policy-resources-templates-policies.yaml.patch]: Resolving - solved with latest commit

**Comment by mihaiplesa** [comment]
@bsclifton what would be needed for changing the cloud.google.com URLs to ours?

**Comment by bsclifton** [comment]
@mihaiplesa I'll check that out next

**Comment by bsclifton** [comment]
@mihaiplesa I found the following places of interest:

1. https://source.chromium.org/chromium/chromium/src/+/main:components/policy/tools/template_writers/writer_configuration.py;l=31;drc=1f2feb50362f0eda403283200fc2687f1150a753

This is one place to make the change. We'd want to have a similar URL / path before we changed it. Something like `https://enterprise.brave.com/policies/`. We might be able to upload the HTML that gets output in `policy_templates.zip`. Would be good to check that out

2. https://source.chromium.org/chromium/chromium/src/+/main:components/policy/resources/webui/policy_base.ts;l=174-187;drc=0cf150a7e8038b9271462fd00de5f25287cc4ccf

This seems to be the webUI - basically if you navigate to brave://policy


I think we'd need to make the change in both places. Could we do that as a follow up?

**Review by kdenhartog — APPROVED** [review]
approving for js-deps-reviewers. No concerns with the changes to package.json.

Some other non-blocking questions that I think are worth considering/following up on:

1. Do we want to add a policy for webTorrent and playlists (could be used to download internal videos)?
2. Would it make sense to follow up with a group policy that configures controls over default lists/shields configurations?
3. Do we have a support doc already written about how to configure and use this? I could see it being a useful link to supply to people who complain about having lots of features they don't want on (tell them how to configure a group policy for themselves by linking to the docs for this page)

**Comment by bsclifton** [comment]
@kdenhartog answers to questions 😄 

1. We could definitely add those types of group policies (web torrent, playlist). This PR is converting the ones we already have into a format that Chromium is using for their policies. Changing the way we format existing ones puts the right stuff into the generated assets (ex: Brave specific items now show - where they didn't before)
2. We do have an existing group policy that can force shields up or down on a list of domains. We don't have a policy for filter lists - that is an interesting one. We could create an issue for this
3. Here's what we have so far: https://support.brave.com/hc/en-us/articles/360039248271-Group-Policy
  Basically, this captures the Brave specific ones we support now. It also has a `policy_templates.zip` download.. but unfortunately, that only has the Chromium policies and not the Brave ones. We do show how to do an example on Windows, macOS, and Linux on the page.  Once this PR is merged, that `policy_templates.zip` will start including the Brave specific ones 😄 We can also update the help article text to mention this. I need to update anyways as SyncUrl was a recent one added which isn't captured yet

File: DEPS [hunk 3]
```diff
@@ -176,6 +176,11 @@ hooks = [
     "condition": 'checkout_mac and host_os != "mac" and checkout_dmg_tool',
     'action': ['build/mac/cross-compile/build-libdmg-hfsplus.py', 'third_party/libdmg-hfsplus']
   },
+  {
+    'name': 'update_group_policy',
+    'pattern': '.',
+    'action': ['python3', 'build/util/update_group_policy.py']
```
Comment (goodov) [inline DEPS:182]: this is not great. A developer expects that modifying a source/policy file will trigger a rebuild. This won't happen in the current approach. Consider doing the copying during the build process, not during the sync.

you can do this by overriding `policy_templates.py` `main()` function.

File: build/util/update_group_policy.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+# copy the group policy files from:
+# `src/brave/components/policy/resources/templates/policy_definitions/`
+# to their expected place in Chromium:
+# `src/components/policy/resources/templates/policy_definitions/`
+#
+
+
+def update_policy_files():
+    policy_dir = wspath(
+        "components/policy/resources/templates/policy_definitions/")
```
Comment (goodov) [inline build/util/update_group_policy.py:34]: pls use root-based paths to avoid confusion. For the reader it's unclear whether this is relative to the script location or to `//brave`. I assume this should be `//brave/components/...`

File: components/policy/resources/templates/policy_definitions/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+2. Update the properties in that file accordingly. Look at some of the existing ones as an example.
+3. Go into `src/brave/browser/policy/brave_simple_policy_map.h` and add your entry here. It'll be auto-generated as `policy::key::k` and then the policy name. With the above example, that would be `policy::key::kBraveRewardsDisable`. You must map this to a profile preference. That is what you'll check in the code.
+4. Go into `src/brave/chromium_src/components/policy/resources/policy_templates.py` and add your policy name to the end of the existing string array there.
+5. In the code where you want to check the profile preference, you can tell if it's set via policy by checking `prefs->IsManagedPreference(<profile_preference_key_here>)`. If this is set to true, you might want to have the UI display something like "This setting is managed by your organization" and have it be read-only.
+6. Build the project like regular (`npm run build`)
```
Comment (goodov) [inline components/policy/resources/templates/policy_definitions/README.md:16]: see, the build instruction is already wrong, because you don't mention that you need to run `npm run sync` to copy new policies to the upstream dir.

pls try the `main()` override in `policy_templates.py` and see how it works.

File: chromium_src/components/policy/tools/generate_policy_source.py [hunk 6]
```diff
@@ -3,198 +3,7 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this file,
 # You can obtain one at https://mozilla.org/MPL/2.0/.
 
-import override_utils
-
+# This override controls the constant written out to:
+# `src/out/<build_type_here>/gen/components/policy/policy_constants.cc`
+# which is then used for the `policy_templates.zip`
 CHROMIUM_POLICY_KEY = 'SOFTWARE\\\\Policies\\\\BraveSoftware\\\\Brave'
```
Comment (goodov) [inline chromium_src/components/policy/tools/generate_policy_source.py:9]: let's add `assert('CHROMIUM_POLICY_KEY' in globals())`

File: chromium_src/components/policy/resources/policy_templates.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+    # append our entries to the ones from policies.yaml
+    # TODO(bsclifton): we can create this array dynamically by reading the file
+    # names from:
+    # `src/brave/components/policy/resources/templates/policy_definitions/BraveSoftware` # pylint: disable=line-too-long
+    brave_policies = [
+        'TorDisabled', 'BraveRewardsDisabled', 'BraveWalletDisabled',
+        'BraveVPNDisabled', 'BraveAIChatEnabled', 'BraveSyncUrl',
+        'BraveShieldsDisabledForUrls', 'BraveShieldsEnabledForUrls'
+    ]
```
Comment (goodov) [inline chromium_src/components/policy/resources/policy_templates.py:62]: if you move the copying of the files into this script, then you don't need to list those here manually.

File: package.json [hunk 8]
```diff
@@ -46,7 +46,8 @@
     "gen_env": "node ./build/commands/scripts/genEnv.js",
     "gen_gradle": "node ./build/commands/scripts/commands.js gen_gradle",
     "ios_pack_js": "webpack --config ios/brave-ios/webpack.config.js",
-    "ios_bootstrap": "node ./build/commands/scripts/iosCommands.js ios_bootstrap"
+    "ios_bootstrap": "node ./build/commands/scripts/iosCommands.js ios_bootstrap",
+    "build_policy_templates": "node ./build/commands/scripts/commands.js build --target=brave_group_policy_templates"
```
Comment (goodov) [inline package.json:50]: 1. please avoid adding "shortcut" commands for cases like this. You can just call `npm run build -- --target=brave_group_policy_templates` where you need to. It's fine to have a separate `npm run` command if we add some custom logic in `build/commands/**/*.js`, but here we're just calling a GN target.

2. consider renaming the GN target to have a verb, for ex.: `generate_brave_group_policy_templates`.

File: components/policy/resources/templates/policy_definitions/README.md [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+These are Brave specific group policies.
+
+Our official documentation is [available here](https://support.brave.com/hc/en-us/articles/360039248271-Group-Policy).
+
+For information about adding a new policy, please see [this page](https://source.chromium.org/chromium/chromium/src/+/main:docs/enterprise/add_new_policy.md).
+
+In a nutshell, the steps for adding a new policy in Brave look like this:
+
+1. Create a new .yaml file under `src/brave/components/policy/resources/templates/policy_definitions/BraveSoftware/`. The name of the file itself will be the policy name. Chromium uses capital casing. For example, we have a policy for disabling Brave Rewards called `BraveRewardsDisable.yaml`. The name used for matching is `BraveRewardsDisable`.
```
Comment (goodov) [inline components/policy/resources/templates/policy_definitions/README.md:11]: prefer using workspace-based paths, like `//brave/components/...`.

File: build/util/update_group_policy.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+def copy_only_if_modified(src, dst):
+    """Copy file if it doesn't exist or if its hash is different."""
+
+    def file_hash(file_path):
+        with open(file_path, "rb") as f:
+            return hashlib.file_digest(f, "sha256").digest()
+
+    if not os.path.exists(dst) or file_hash(src) != file_hash(dst):
+        shutil.copy2(src, dst)
```
Comment (goodov) [inline build/util/update_group_policy.py:58]: please also check that REMOVING a file from the the brave directory won't trigger a build error, because in the current approach we only copy/replace files, but never remove them from the upstream directory.

File: package.json (hunk 8 above)
Comment (mherrmann) [reply package.json:50]: I don't have a strong opinion on 2. but [concur](https://github.com/brave/brave-core/pull/25699#discussion_r1773575641) with 1.

File: DEPS (hunk 3 above)
Comment (mherrmann) [reply DEPS:182]: I concur with this too.

File: BUILD.gn [hunk 11]
```diff
@@ -247,6 +247,26 @@ if (!is_ios) {
   }
 }
 
+if (!is_android && !is_ios) {
```
Comment (mherrmann) [inline BUILD.gn:250]: I think we should just replace `!is_android && !is_ios` by `is_win`.

File: build/util/update_group_policy.py (hunk 4 above)
Comment (bsclifton) [reply build/util/update_group_policy.py:34]: I didn't know we can use that style of pathing here - I'll give it a shot

File: BUILD.gn (hunk 11 above)
Comment (bsclifton) [reply BUILD.gn:250]: I might remove this - thanks @goodov for the feedback about avoiding convenience aliases. You had mentioned that too @mherrmann. I'll get that fixed up and I think we can remove this section

File: components/policy/resources/templates/policy_definitions/BraveSoftware/BraveAIChatEnabled.yaml [hunk 12]
```diff
@@ -0,0 +1,39 @@
+caption: Disable AI Chat
+default: null
+desc: |-
+  Disable AI Chat in Brave.
+
+  The smart AI assistant built right into your browser. Ask questions, summarize pages, create new content, and more. Privately.
```
Comment (bsclifton) [inline components/policy/resources/templates/policy_definitions/BraveSoftware/BraveAIChatEnabled.yaml:6]: cc: @mattmcalister on the wording here! I grabbed it from https://brave.com/leo/. This text will show in the Windows group policy editor. It looks like this when it's displayed:
image

File: components/policy/resources/templates/policy_definitions/BraveSoftware/BraveRewardsDisabled.yaml [hunk 13]
```diff
@@ -0,0 +1,37 @@
+caption: Disable Brave Rewards
+default: null
+desc: |-
+  Disable Brave Rewards in Brave.
+
+  Allows users to support content creators and earn tokens for ads seen in Brave. Folks can use tokens to buy gift cards, exchange for other currencies, and more..
```
Comment (bsclifton) [inline components/policy/resources/templates/policy_definitions/BraveSoftware/BraveRewardsDisabled.yaml:6]: cc: @Miyayes on the wording here! I grabbed it from https://brave.com/brave-rewards/. This text will show in the Windows group policy editor. It looks like this when it's displayed (this screenshot is for Wallet, but just imagine the text above being show for the Rewards group policy)
image

File: components/policy/resources/templates/policy_definitions/BraveSoftware/BraveWalletDisabled.yaml [hunk 14]
```diff
@@ -0,0 +1,37 @@
+caption: Disable Brave Wallet
+default: null
+desc: |-
+  Disable Brave Wallet in Brave.
+
+  Allows users to use the secure, built-in crypto wallet that supercharges the browser for Web3.
```
Comment (bsclifton) [inline components/policy/resources/templates/policy_definitions/BraveSoftware/BraveWalletDisabled.yaml:6]: cc: @Miyayes on the wording here! I grabbed it from https://brave.com/wallet/. This text will show in the Windows group policy editor. It looks like this when it's displayed
image

File: components/policy/resources/templates/policy_definitions/BraveSoftware/TorDisabled.yaml [hunk 15]
```diff
@@ -0,0 +1,39 @@
+caption: Disable Tor functionality
+default: null
+desc: |-
+  Disable Tor functionality in Brave.
+
+  Using Tor makes it more difficult to trace a user's Internet activity by preventing any single point on the Internet (other than the user's device) from being able to view both where traffic originated from and where it is ultimately going to at the same time. This conceals a user's location and usage from anyone performing network surveillance or traffic analysis from any such point, protecting the user's freedom and ability to communicate confidentially.
```
Comment (bsclifton) [inline components/policy/resources/templates/policy_definitions/BraveSoftware/TorDisabled.yaml:6]: cc: @diracdeltas @fmarier on the wording here! I copied it from Wikipedia 😱 Maybe we have a better summary that we should show. It shows in the Windows group policy editor which looks like this:
image

File: components/policy/resources/templates/policy_definitions/BraveSoftware/BraveVPNDisabled.yaml [hunk 16]
```diff
@@ -0,0 +1,38 @@
+caption: Disable Brave VPN
+default: null
+desc: |-
+  Disable Brave VPN in Brave.
+
+  Allows users to use Brave VPN to block trackers and encrypt every connection to the Web, on every app, even outside the Brave browser.
```
Comment (bsclifton) [inline components/policy/resources/templates/policy_definitions/BraveSoftware/BraveVPNDisabled.yaml:6]: cc: @mattmcalister on the wording here! I copied it from https://brave.com/firewall-vpn/. It'll show in the Windows group policy editor which looks like this:
image

File: package.json (hunk 8 above)
Comment (bsclifton) [reply package.json]: Removed! Thanks for the feedback - @mherrmann had raised. I went ahead and removed. We can continue to use `npm run build -- --target=brave/components/policy:pack_policy_templates`

File: build/util/update_group_policy.py (hunk 4 above)
Comment (bsclifton) [reply build/util/update_group_policy.py]: done

File: components/policy/resources/templates/policy_definitions/README.md (hunk 9 above)
Comment (bsclifton) [reply components/policy/resources/templates/policy_definitions/README.md]: Fixed for everything 😄👍

File: DEPS (hunk 3 above)
Comment (bsclifton) [reply DEPS]: OK good call - done. Need to try it out 😄

File: components/policy/resources/templates/policy_definitions/BraveSoftware/TorDisabled.yaml (hunk 15 above)
Comment (fmarier) [reply components/policy/resources/templates/policy_definitions/BraveSoftware/TorDisabled.yaml:6]: Seems good to me, but @diracdeltas may be aware of past discussions we've had with the Tor team about how we describe what Tor does.

File: components/policy/resources/templates/policy_definitions/README.md (hunk 5 above)
Comment (bsclifton) [reply components/policy/resources/templates/policy_definitions/README.md:16]: This approach seems to be working great! I'll wrap up testing tomorrow (for me; Thursday) and update this documentation

File: chromium_src/components/policy/resources/policy_templates.py [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+                continue
+            src_dir = entry.path
+            src_dir_name = entry.name
+            dst_dir = wspath(
+                f"//components/policy/resources/templates/policy_definitions/{src_dir_name}"  # pylint: disable=line-too-long
+            )
+            shutil.copytree(src_dir,
+                            dst_dir,
+                            dirs_exist_ok=True,
+                            copy_function=copy_only_if_modified)
```
Comment (mherrmann) [inline chromium_src/components/policy/resources/policy_templates.py:107]:
```suggestion
    with os.scandir(policy_dir) as entries:
        for entry in entries:
            if not entry.is_dir():
                continue
            dst_dir = wspath(
                f"//components/policy/resources/templates/policy_definitions/" +
                entry.name
            )
            shutil.copytree(entry.path,
                            dst_dir,
                            dirs_exist_ok=True,
                            copy_function=copy_only_if_modified)
```

Comment (mherrmann) [reply chromium_src/components/policy/resources/policy_templates.py:107]: Also, I believe the present implementation does not handle the case where one of our `.yaml` policy files gets deleted.

File: components/policy/resources/templates/policy_definitions/README.md [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+
+For information about adding a new policy, please see [this page](https://source.chromium.org/chromium/chromium/src/+/main:docs/enterprise/add_new_policy.md).
+
+In a nutshell, the steps for adding a new policy in Brave look like this:
+
+1. Create a new .yaml file under `//brave/components/policy/resources/templates/policy_definitions/BraveSoftware/`. The name of the file itself will be the policy name. Chromium uses capital casing. For example, we have a policy for disabling Brave Rewards called `BraveRewardsDisable.yaml`. The name used for matching is `BraveRewardsDisable`.
+2. Update the properties in that file accordingly. Look at some of the existing ones as an example.
+3. Go into `//brave/browser/policy/brave_simple_policy_map.h` and add your entry here. It'll be auto-generated as `policy::key::k` and then the policy name. With the above example, that would be `policy::key::kBraveRewardsDisable`. You must map this to a profile preference. That is what you'll check in the code.
+4. Go into `//brave/chromium_src/components/policy/resources/policy_templates.py` and add your policy name to the end of the existing string array there.
```
Comment (mherrmann) [inline components/policy/resources/templates/policy_definitions/README.md:14]: I suspect this point 4. will become obsolete when you read the files dynamically.

File: components/policy/resources/templates/policy_definitions/README.md [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+5. In the code where you want to check the profile preference, you can tell if it's set via policy by checking `prefs->IsManagedPreference(<profile_preference_key_here>)`. If this is set to true, you might want to have the UI display something like "This setting is managed by your organization" and have it be read-only.
+6. Build the project like regular (`npm run build`)
+7. You can build the policy templates using `npm run build_policy_templates`.
```
Comment (mherrmann) [inline components/policy/resources/templates/policy_definitions/README.md:17]: This is obsolete I think.

File: components/policy/resources/templates/policy_definitions/README.md [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+8. The group policy assets get output to `//out/<build_type_here>/brave_policy_templates.zip`. You can open this up and inspect the registry and adm/admx files.
+
+If rebuilding isn't picking up your policy, you can try to delete `//out/<build_type_here>/gen/components/policy/policy_constants.cc`. Rebuilding will trigger the rebuild for this file and your charges should be there.
```
Comment (mherrmann) [inline components/policy/resources/templates/policy_definitions/README.md:20]:
```suggestion
If rebuilding isn't picking up your policy, you can try to delete `//out/<build_type_here>/gen/components/policy/policy_constants.cc`. Rebuilding will trigger the rebuild for this file and your changes should be there.
```

File: components/policy/resources/templates/policy_definitions/README.md [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+
+Note that the hooks (ex: ran when you do `npm run sync`) are what copy the Brave policies to the `//components/policy/..` directory structure. If you make a change to the policy files, you'll need to either manually copy the file(s) or run the hooks again.
```
Comment (mherrmann) [inline components/policy/resources/templates/policy_definitions/README.md:22]: Obsolete.

Comment (bsclifton) [reply components/policy/resources/templates/policy_definitions/README.md]: Fixed! Great catch, thank you! 😄

File: components/policy/resources/templates/policy_definitions/README.md (hunk 20 above)
Comment (bsclifton) [reply components/policy/resources/templates/policy_definitions/README.md]: Fixed 😄

File: components/policy/resources/templates/policy_definitions/README.md (hunk 19 above)
Comment (bsclifton) [reply components/policy/resources/templates/policy_definitions/README.md]: Updated to the original way 😄 Thanks!

File: chromium_src/components/policy/resources/policy_templates.py [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+    # "policies"
+    # This has the contents of:
+    # `//components/policy/resources/templates/policies.yaml`
+    # This is where we need to inject the Brave specific names. The policies
+    # themselves are already defined (under `policy_definitions`), we just need
+    # to add a mapping for ID (integer; unique) and name (matches name under
+    # `policy_definitions`).
+    #
+    #
+    # There are some other fields which are not used by this script:
```
[the remark this answers is not in the document]
Comment (bsclifton) [reply chromium_src/components/policy/resources/policy_templates.py]: Fixed!

File: chromium_src/components/policy/resources/policy_templates.py [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+    #
+    # "policy_definitions"
+    # there will be one "group" for every folder found under
+    # `//components/policy/resources/templates/policy_definitions`
+    # Chromium considers the folder name the group name for the policy.
+    # Brave uses the group name "BraveSoftware". The child element for the
+    # group is the policy itself (those are the yaml files in the folder).
+    #
+    # Brave specific entries will get copied into place by the hooks that run
+    # via DEPS. We copy the files from:
```
[the remark this answers is not in the document]
Comment (bsclifton) [reply chromium_src/components/policy/resources/policy_templates.py]: Fixed 😄

File: build/util/update_group_policy.py (hunk 10 above)
Comment (bsclifton) [reply build/util/update_group_policy.py]: Definitely no problem here - I did test that. Basically, it adds a constant into `//out/<build_type_here>/gen/components/policy/policy_constants.cc` which would then just never get used

File: components/policy/resources/templates/policy_definitions/README.md (hunk 18 above)
Comment (bsclifton) [reply components/policy/resources/templates/policy_definitions/README.md:14]: OK I think this should be sorted; I'll test the script and then remove this section 😄

File: chromium_src/components/policy/resources/policy_templates.py (hunk 7 above)
Comment (bsclifton) [reply chromium_src/components/policy/resources/policy_templates.py]: Fixed. No need to examine the files during copy (although that was great to combine the two). The other field policy_definitions had all the items all along 😄 Updated!

**Comment by mihaiplesa** [comment]
@bsclifton what URL is there in the description of each policy now?

File: components/policy/resources/templates/policy_definitions/README.md [hunk 24]
```diff
@@ -0,0 +1,20 @@
...
+3. Go into `//brave/browser/policy/brave_simple_policy_map.h` and add your entry here. It'll be auto-generated as `policy::key::k` and then the policy name. With the above example, that would be `policy::key::kBraveRewardsDisable`. You must map this to a profile preference (you must create a new one). That new preference is what you'll check in the code.
+4. Go into `//brave/chromium_src/components/policy/resources/policy_templates.py` and add your policy name to the end of the existing string array there.
+5. In the code where you want to check the profile preference, you can tell if it's set via policy by checking `prefs->IsManagedPreference(<profile_preference_key_here>)`. If this is set to true, you might want to have the UI display something like "This setting is managed by your organization" and have it be read-only.
+6. Build the project like regular (`npm run build`).
+7. You can build the policy templates using `npm run build -- --target=brave/components/policy:pack_policy_templates`.
+8. The group policy assets get output to `//out/<build_type_here>/brave_policy_templates.zip`. You can open this up and inspect the registry and adm/admx files.
```
Comment (mihaiplesa) [inline components/policy/resources/templates/policy_definitions/README.md:18]: Is this still true? Or it is `policy_templates.zip`?

File: components/policy/resources/templates/policy_definitions/BraveSoftware/BraveRewardsDisabled.yaml (hunk 13 above)
Comment (Miyayes) [reply components/policy/resources/templates/policy_definitions/BraveSoftware/BraveRewardsDisabled.yaml:6]: Brave Rewards is a feature that allows users to get paid for ads they see in Brave and send contributions to their favorite websites and content creators.

File: components/policy/resources/templates/policy_definitions/README.md (hunk 24 above)
Comment (bsclifton) [reply components/policy/resources/templates/policy_definitions/README.md:18]: Yup, this is true. Folks can use the `policy_templates.zip` - but those haven't had the script that renames Chrome => Brave run yet. The nice thing about `policy_templates.zip` is that it also has HTML documentation.

But yeah - we are currently sharing the `brave_policy_templates.zip` on the support page (just renamed as `policy_templates.zip`). Unless we make a change, that's the one we'll want to continue using

Comment (bsclifton) [reply components/policy/resources/templates/policy_definitions/README.md]: Updated the README! 😄

**Comment by bsclifton** [comment]
> @bsclifton what URL is there in the description of each policy now?
I didn't change that yet. Want me to change that as part of this same PR too? I think I can hardcode it as https://support.brave.com/hc/en-us/articles/360039248271-Group-Policy

File: components/policy/resources/templates/policy_definitions/README.md (hunk 18 above)
Comment (bsclifton) [reply components/policy/resources/templates/policy_definitions/README.md]: Removed 👍

**Comment by mihaiplesa** [comment]
@bsclifton hardcoded is fine as long as we keep the `?policy=xxx` at the end of each URL.

**Review by mherrmann — APPROVED** [review]

**Comment by bsclifton** [comment]
OK I have found how to change the URL (changed in last commit). But I would like to investigate a better way to do this (without a patch, if possible).
image

**Comment by bsclifton** [comment]
I've tested this out and it's working great - as long as we're OK with the patch done for the URL.

Will let reviewers get a chance to re-review (@goodov or maybe @bridiver)

**Comment by kdenhartog** [comment]
Responding to @bsclifton from https://github.com/brave/brave-core/pull/25710#issuecomment-2387655634

> We could definitely add those types of group policies (web torrent, playlist). This PR is converting the ones we already have into a format that Chromium is using for their policies. Changing the way we format existing ones puts the right stuff into the generated assets (ex: Brave specific items now show - where they didn't before)

Sounds good, I'll work on opening separate issues for these and labeling them as good first issues. They likely are ones that can be handled by community members to help contribute to the project.

> We do have an existing group policy that can force shields up or down on a list of domains. We don't have a policy for filter lists - that is an interesting one. We could create an issue for this

I'll open a separate issue for this too and label it the same as the ones above too.

> Feedback welcome! Basically, this captures the Brave specific ones we support now. It also has a policy_templates.zip download.. but unfortunately, that only has the Chromium policies and not the Brave ones. We do show how to do an example on Windows, macOS, and Linux on the page. Once this PR is merged, that policy_templates.zip will start including the Brave specific ones 😄 We can also update the help article text to mention this. I need to update anyways as SyncUrl was a recent one added which isn't captured yet

The description of how this work makes sense to me as I was reading it. When I originally asked the question I wasn't even sure if we had docs on it which this answered the question about. The only piece of feedback I came up with was to maybe address that critic audience better, but I'm not certain. For now, as I come across these people I'll start linking to this doc and letting them know they can easily un-configure these features using the group policy and see how they respond before actually trying to solve a problem that may not be a problem at all (e.g. they may just be complaining for the sake of it, not because this is an actual pain point).

**Review by goodov — APPROVED** [review]
lgtm with few things to improve

File: chromium_src/components/policy/resources/policy_templates.py [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+    # `//brave/browser/policy/brave_simple_policy_map.h`
+    #
+    # When the code below is ran this will copy the group policy files from:
+    # `//brave/components/policy/resources/templates/policy_definitions/`
+    # to their expected place in Chromium:
+    # `//components/policy/resources/templates/policy_definitions/`
+    #
+    policy_dir = wspath(
+        "//brave/components/policy/resources/templates/policy_definitions/")
+    with os.scandir(policy_dir) as entries:
```
[the remark this answers is not in the document]
Comment (bsclifton) [reply chromium_src/components/policy/resources/policy_templates.py:79]: no we don't - let me remove. Thanks! 😄

File: components/policy/resources/templates/policy_definitions/BraveSoftware/TorDisabled.yaml (hunk 15 above)
Comment (diracdeltas) [reply components/policy/resources/templates/policy_definitions/BraveSoftware/TorDisabled.yaml:6]: I would just copy the text we already have in Private Tabs; change title to "Disable Tor Connectivity" and change the text to:

Disable Tor connectivity in Brave.

Using Tor connectivity in Private Windows makes it more difficult for sites to see a user's true IP address and for network observers to see what sites they visit. This may be used for enhanced confidentiality and censorship resistance in Private Windows.

If this policy is set to true, Tor connectivity will always be disabled.

If this policy is set to false, Tor connectivity will be enabled if the user opens a Private Window with Tor. 

If you set this policy, users cannot change or override it.

If this policy is left unset, Tor connectivity will be enabled if the user opens a Private Window with Tor.

Comment (bsclifton) [reply components/policy/resources/templates/policy_definitions/BraveSoftware/TorDisabled.yaml]: Updated! 😄 Thanks

**Comment by bsclifton** [comment]
Made final changes - thanks @goodov, @mherrmann, and @diracdeltas and everyone who helped review 😄👍
